src/Cli/dotnet/commands/dotnet-workload/restore/WorkloadRestoreCommand.cs

```diff
@@ -2,43 +2,119 @@
 using Microsoft.DotNet.Cli;
 using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Workloads.Workload.Install;
 using Microsoft.Extensions.EnvironmentAbstractions;
+using Microsoft.NET.Sdk.WorkloadManifestReader;
 
 namespace Microsoft.DotNet.Workloads.Workload.Restore
 {
     internal class WorkloadRestoreCommand : CommandBase
     {
-        private readonly string _configFilePath;
-        private readonly IReporter _errorReporter;
-        private readonly IFileSystem _fileSystem;
+        private readonly ParseResult _result;
         private readonly IReporter _reporter;
-        private readonly string[] _sources;
-        private readonly string _verbosity;
+        private readonly IEnumerable<string> _slnOrProjectArgument;
 
         public WorkloadRestoreCommand(
             ParseResult result,
             IFileSystem fileSystem = null,
             IReporter reporter = null)
             : base(result)
         {
-            _fileSystem = fileSystem ?? new FileSystemWrapper();
-
+            _result = result;
             _reporter = reporter ?? Reporter.Output;
-            _errorReporter = reporter ?? Reporter.Error;
-
-            _configFilePath = result.ValueForOption<string>(WorkloadRestoreCommandParser.ConfigOption);
-            _sources = result.ValueForOption<string[]>(WorkloadRestoreCommandParser.SourceOption);
-            _verbosity =
-                Enum.GetName(result.ValueForOption<VerbosityOptions>(WorkloadRestoreCommandParser.VerbosityOption));
+            _slnOrProjectArgument =
+                result.ValueForArgument<IEnumerable<string>>(RestoreCommandParser.SlnOrProjectArgument);
         }
 
         public override int Execute()
         {
-            _reporter.WriteLine("WIP workload restore stub");
+            var allProjects = DiscoverAllProjects(Directory.GetCurrentDirectory(), _slnOrProjectArgument).Distinct();
+
+            List<WorkloadId> allWorkloadId = RunTargetToGetWorkloadIds(allProjects);
+
+            _reporter.WriteLine(string.Format(LocalizableStrings.InstallingWorkloads, string.Join(" ", allWorkloadId)));
+
+            var workloadInstallCommand = new WorkloadInstallCommand(_result,
+                workloadIds: new ReadOnlyCollection<string>(allWorkloadId.Select(a => a.ToString()).ToList()));
+
+            workloadInstallCommand.Execute();
             return 0;
         }
+
+        private static List<WorkloadId> RunTargetToGetWorkloadIds(IEnumerable<string> allProjects)
+        {
+            var globalProperties = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
+            {
+                {Constants.MSBuildExtensionsPath, AppContext.BaseDirectory}
+            };
+
+            var allWorkloadId = new List<WorkloadId>();
+            foreach (string projectFile in allProjects)
+            {
+                var project = new ProjectInstance(projectFile, globalProperties, null);
+
+                project.Build(new[] {"_FindAllReferenceWorkload"},
```

Comment:
@rainersigwald please help me double check that's all I need for using msbuild API calling a target

Reply:
Can you check for success?

Reply:
You could probably improve perf here by doing a `BeginBuild`/[PendBuildRequest](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.execution.buildmanager.pendbuildrequest?view=msbuild-16-netcore)/`EndBuild` flow so MSBuild can deduplicate some work and do a bunch of stuff concurrently. Not critical for this operation IMO.